src/cdrizzleapi.c

```diff
@@ -48,16 +48,160 @@ scale_image(PyArrayObject *image, double scale_factor) {
+        *ny = (int)ndim[0];
+        *arrays = arr_list;
+        return 0;
+
+    } else if (!PyList_Check(list) && !PyTuple_Check(list)) {
+        driz_error_set(
+            error, PyExc_TypeError,
+            "Argument '%s' is not a list or a tuple of numpy.ndarray or None.",
+            name);
+    }
+
+    at_least_one = 0;
+
+    if (!(*nmax = PySequence_Size(list))) {
+        return 0;
+    }
+
+    arr_list = (PyArrayObject **)calloc(*nmax, sizeof(PyArrayObject *));
+    if (!arr_list) {
+        driz_error_set(error, PyExc_MemoryError, "Memory allocation failed.");
+        return 1;
+    }
+
+    for (i = 0; i < *nmax; ++i) {
+        if (!(list_elem = PySequence_GetItem(list, i))) {
+            driz_error_set(error, PyExc_RuntimeError,
+                           "Error retrieving array %d from the '%s' list.", i,
+                           name);
+            goto _exit_on_err;
+        }
+
+        if (list_elem == Py_None) {
+            if (allow_none) {
+                if (n_none) {
+                    (*n_none)++;
+                }
+                arr_list[i] = NULL;
+                Py_XDECREF(list_elem);
+                continue;
+            } else {
+                *n_none = 1;
+                driz_error_set(
+                    error, PyExc_ValueError,
+                    "Element %d of '%s' list is None which is not allowed.", i,
+                    name);
+                goto _exit_on_err;
+            }
+        } else {
+            arr = (PyArrayObject *)PyArray_ContiguousFromAny(list_elem,
+                                                             NPY_FLOAT, 2, 2);
+            if (!arr) {
+                driz_error_set(error, PyExc_ValueError,
+                               "Invalid array in '%s' at position %d.", name,
+                               i);
+                goto _exit_on_err;
+            }
+            at_least_one = 1;
+        }
+        Py_XDECREF(list_elem);
+
+        arr_list[i] = arr;
+
+        if (*nx < 0) {
+            ndim = PyArray_DIMS(arr);
+            *nx = (int)ndim[1];
+            *ny = (int)ndim[0];
+        } else {
+            ndim = PyArray_DIMS(arr);
+            inx = (int)ndim[1];
+            iny = (int)ndim[0];
+            if ((*nx != inx) || (*ny != iny)) {
+                driz_error_set(
+                    error, PyExc_ValueError,
+                    "Inconsistent image shape in the '%s' image list.", name);
+                goto _exit_on_err;
+            }
+        }
+    }
+
+    if (!at_least_one && !allow_none) {
+        free(arr_list);
+        arr_list = NULL;
+        *nmax = 0;
+    }
+
+    *arrays = arr_list;
+    return 0;
+
+_exit_on_err:
+
+    Py_XDECREF(list_elem);
+    if (arr_list) {
+        for (i = 0; i < *nmax; ++i) {
+            Py_XDECREF(arr_list[i]);
+        }
+        free(arr_list);
+    }
+}
+
 static PyObject *
 tdriz(PyObject *obj UNUSED_PARAM, PyObject *args, PyObject *keywords) {
```

Comment:
This function roughly 400 lines long and should be broken down into smaller sub-functions.